### ISSUE TITLE
Drawing on swappy offsets the brush position

### ISSUE THREAD

**Issue by acoolstraw** [labwc/labwc#340]
> * [ ]  Drawing on swappy (`grim -g "$(slurp)" - | swappy -f -`) is buggy, as for some reason the brush poiner is offset diagonally, but your actual cursor stays where you originally started the drawing

Caused by 

https://github.com/labwc/labwc/blob/fbcdf38b52bb5a9635962e346c4f82e8c932ec89/src/cursor.c#L296

The offset is because `scene_xdg_surface_update_position()` takes into account the invisible border of the view and positions the node accordingly in https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/scene/xdg_shell.c#L57

_Originally posted by @johanmalm in https://github.com/labwc/labwc/issues/332#issuecomment-1133950545_

**Comment by Consolatis** [comment]
There are two aspects to this bug:
- it takes the wrong branch that is intended for the `out-of-view-thingie`
- in that branch the calculation seems to ignore invisible border for CSD

Debugging the first aspect:
- checking for `view != seat.active_view` indeed solves the offset
- but breaks the `out-of-view-thingie` (patent pending)
- that in turn is caused by clearing the focus
- which in turn is caused by moving over the SSD (because the view is still the same but there is no surface)
- which can be fixed by checking for `!seat.active_view` in the clear focus banch

I have a patch in the queue and will post a PR later.

The second aspect is still there but not that urgent because absolute coordinates don't really matter all that much while selecting text or dragging a scrollbar, its more about the relative motion in those cases.
That should still be fixed eventually though.

**Comment by Consolatis** [comment]
> in that branch the calculation seems to ignore invisible border for CSD

Swappy is pretty awesome to test the offset (created with the PR applied):
![deco](https://user-images.githubusercontent.com/35009135/170045615-7ab0458f-d0eb-4852-9543-fdcaa36a7b1f.png)

So yeah.. there is definitely some offset missing. With the PR applied one can actually draw fine right and below of the red line as long as the cursor doesn't leave the surface so its really just  an issue in the `out-of-view-thingie`.

**Comment by Consolatis** [comment]
```diff
diff --git a/src/cursor.c b/src/cursor.c
index 81d527a..fe8afab 100644
--- a/src/cursor.c
+++ b/src/cursor.c
@@ -311,6 +311,12 @@ process_cursor_motion(struct server *server, uint32_t time)
                sy = server->seat.cursor->y - view->y;
                sx = sx < 0 ? 0 : (sx > view->w ? view->w : sx);
                sy = sy < 0 ? 0 : (sy > view->h ? view->h : sy);
+               if (view->type == LAB_XDG_SHELL_VIEW && view->xdg_surface) {
+                       struct wlr_box geo;
+                       wlr_xdg_surface_get_geometry(view->xdg_surface, &geo);
+                       sx += geo.x;
+                       sy += geo.y;
+               }
                wlr_seat_pointer_notify_motion(server->seat.seat, time, sx, sy);
        } else if (surface && !input_inhibit_blocks_surface(
                        &server->seat, surface->resource)) {
```
This fixes the offset for wayland native `out-of-view-with-mouse-button-pressed` use cases.. but I am really not sure we are at the right area for this fix.

**Comment by johanmalm** [comment]
Yes, doesn't feel quite right because with swaggy we're `within-view-with-mouse-button-pressed`
